src/library/searchquery.cpp

```diff
@@ -471,6 +471,217 @@ double DurationFilterNode::parse(const QString& arg, bool* ok) {
     return 60 * m + s;
 }
 
+// static
+constexpr double BpmFilterNode::kRelativeRangeDefault;
+
+// static
+double BpmFilterNode::s_relativeRange = kRelativeRangeDefault;
+
+// static
+void BpmFilterNode::setBpmRelativeRange(double range) {
+    // range < 0 would yield zero results because m_dRangeLow > m_dRangeHigh
+    VERIFY_OR_DEBUG_ASSERT(range >= 0) {
+        return;
+    }
+    s_relativeRange = range;
+}
+
+namespace {
+
+inline QString rangeSqlString(double lower, double upper) {
+    return QStringLiteral("bpm BETWEEN %1 AND %2")
+            .arg(QString::number(lower),
+                    QString::number(upper));
+}
+
+inline std::pair<double, double> rangeFromTrailingDecimal(double bpm) {
+    // Set up a range if we have decimals. This will include matches
+    // for which we show rounded values in the library. For example
+    // 124  finds 119.95 - 124.05
+    // 124.1  finds 124.05 - 124.15
+    // 124.92 finds 124.915 - 124.925
+    if (bpm == 0.0) {
+        return std::pair<double, double>(bpm, bpm);
+    }
+    int numDecimals = 1;
+    double intPart;
+    double fractPart = std::modf(bpm, &intPart);
+    if (fractPart != 0.0) {
+        QString decimals = QString::number(fractPart);
+        numDecimals = decimals.split('.').at(1).length();
+    }
+    double roundRange = 5 / pow(10, numDecimals + 1);
+    // Don't search for negative BPM
+    double lower = std::max(0.0, bpm - roundRange);
+    double upper = bpm + roundRange;
+    return std::pair<double, double>(lower, upper);
+}
+
+} // namespace
+
+BpmFilterNode::BpmFilterNode(QString& argument, bool fuzzy, bool negate)
+        : m_matchMode(MatchMode::Invalid),
+          m_operator("="),
+          m_bpm(0.0),
+          m_rangeLower(0.0),
+          m_rangeUpper(0.0),
+          m_bpmHalfLower(0.0),
+          m_bpmHalfUpper(0.0),
+          m_bpmDoubleLower(0.0),
+          m_bpmDoubleUpper(0.0) {
+    if (argument == kMissingFieldSearchTerm) {
+        m_matchMode = MatchMode::Null;
+        return;
+    }
+
+    QRegularExpressionMatch opMatch = kNumericOperatorRegex.match(argument);
+    if (opMatch.hasMatch()) {
+        if (fuzzy) {
+            // fuzzy can't be combined with operators
+            // m_matchMode is already Invalid.
+            return;
+        }
+        m_operator = opMatch.captured(1);
+        argument = opMatch.captured(2);
+    }
+
+    // Replace the locale's decimal separator with .
+    // This is handy if numbers are typed with the numpad.
+    argument.replace(',', '.');
+    bool isDouble = false;
+    double bpm = argument.toDouble(&isDouble);
```

Comment:
bpm is only used after this mode conditional branch, can you move it down to, say, line 600 or so? That way it's closer to where it's being used

Reply:
Well, we need to know if the argument is a double here already, and doing so with toDouble() gives us the bpm.
We may assign m_bpm = bpm right away at the top of the isDouble branch, regardless whether it'll be used or not (that would save the ::Operator case at the end of this function).